server/app/services/applicant/ApplicantService.java

```diff
@@ -1101,4 +1113,128 @@ abstract static class Builder {
+  private ImmutableMap<String, String> buildCorrectedAddressAsFormData(
+      long applicantId,
+      long programId,
+      String blockId,
+      AddressQuestion addressQuestion,
+      Optional<AddressSuggestion> suggestionMaybe,
+      String selectedAddress) {
+
+    Map<String, String> questionPathToValueMap = new HashMap<>(Map.of());
+
+    if (suggestionMaybe.isPresent()) {
+      AddressSuggestion suggestion = suggestionMaybe.get();
+      Address address = suggestion.getAddress();
+      AddressLocation location = suggestion.getLocation();
+
+      questionPathToValueMap.put(addressQuestion.getStreetPath().toString(), address.getStreet());
+      questionPathToValueMap.put(addressQuestion.getLine2Path().toString(), address.getLine2());
+      questionPathToValueMap.put(addressQuestion.getCityPath().toString(), address.getCity());
+      questionPathToValueMap.put(addressQuestion.getStatePath().toString(), address.getState());
+      questionPathToValueMap.put(addressQuestion.getZipPath().toString(), address.getZip());
+      questionPathToValueMap.put(
+          addressQuestion.getLatitudePath().toString(), location.getLatitude().toString());
+      questionPathToValueMap.put(
+          addressQuestion.getLongitudePath().toString(), location.getLongitude().toString());
+      questionPathToValueMap.put(
+          addressQuestion.getWellKnownIdPath().toString(), location.getWellKnownId().toString());
+      questionPathToValueMap.put(
+          addressQuestion.getCorrectedPath().toString(),
+          CorrectedAddressState.CORRECTED.toString());
+    } else if (selectedAddress.equals(AddressCorrectionBlockView.USER_KEEPING_ADDRESS_VALUE)) {
+      questionPathToValueMap.put(
+          addressQuestion.getCorrectedPath().toString(),
+          CorrectedAddressState.AS_ENTERED_BY_USER.toString());
+    } else {
+      questionPathToValueMap.put(
+          addressQuestion.getCorrectedPath().toString(), CorrectedAddressState.FAILED.toString());
+      logger.error(
+          "Address correction failed for applicantId: {} programId: {} blockId: {}",
+          applicantId,
+          programId,
+          blockId);
+    }
+
+    return ImmutableMap.<String, String>builder().putAll(questionPathToValueMap).build();
+  }
+
+  /**
+   * Finds the first {@link ApplicantQuestion} that is an address question type and has address
+   * correction enabled.
+   */
+  public ApplicantQuestion getFirstAddressCorrectionEnabledApplicantQuestion(Block block) {
+    Optional<ApplicantQuestion> applicantQuestionMaybe =
+        block.getAddressQuestionWithCorrectionEnabled();
+
+    if (applicantQuestionMaybe.isEmpty()) {
+      throw new RuntimeException(
+          String.format(
+              "Expected to find an address with address correction enabled in block %s, but did"
+                  + " not.",
+              block.getId()));
+    }
+
+    return applicantQuestionMaybe.get();
+  }
+
+  /** Gets address suggestions */
```

Comment:
**suggestion (blocking):** please expand on this javadoc to mention ESRI and the relevance of the method's parameter

Reply:
done